Topics/Tech_Stacks/Web-Sockets.md

```diff
@@ -0,0 +1,162 @@
+# Building Interactive Web Applications with WebSockets
+
+## Table of Contents
+1. [Introduction](#introduction)
+2. [Understanding WebSockets](#understanding-websockets)
+3. [WebSockets vs HTTP](#websockets-vs-http)
+4. [Advantages and Disadvantages](#advantages-and-disadvantages)
+5. [Setting Up](#setting-up)
+6. [Real-World Use Cases](#real-world-use-cases)
+7. [Conclusion](#conclusion)
+8. [Further Reading](#further-reading)
+
+## Introduction
+In the evolving landscape of web development, the need for real-time, interactive web applications is more prominent than ever. Users expect dynamic experiences with instant updates—be it in chat applications, live sports scores, or stock market feeds. WebSockets represent a significant advancement in web communication, allowing for real-time, full-duplex communication between the client and server. This technology is especially relevant to the CSC301 curriculum as it introduces students to modern web development practices, preparing them for the challenges of developing interactive, user-centric software solutions.
+
+## Understanding WebSockets
+WebSockets are a powerful protocol embedded within the familiar web technology stack, enabling a significant leap in communication capabilities over the Internet. They provide a way to overcome the limitations of the traditional HTTP request-response model, especially when it comes to real-time, low-latency communication.
+
+Unlike HTTP, where a new connection is established for each request and response, WebSockets establish a single, persistent connection between the client and the server. This connection is initiated by the client sending a special type of HTTP request, asking the server to upgrade the connection to a WebSocket connection.
+
+### How the WebSocket Connection is Established:
+1. **Client Handshake Request:** The client initiates the connection by sending a WebSocket handshake request to the server. This request looks like a standard HTTP request but includes an `Upgrade: websocket` header that signals the desire to establish a WebSocket connection.
+   
+2. **Server Handshake Response:** If the server supports WebSockets and accepts the connection request, it responds with a handshake of its own, confirming the upgrade from HTTP to WebSockets.
+   
+3. **Persistent Connection:** With the handshake complete, the connection remains open, and data can now be sent in either direction at any time, allowing for two-way, **real-time communication**.
+
+### WebSocket Communication:
+Once established, the WebSocket connection allows messages to be passed back and forth between the client and server with minimal overhead. Each message is framed in a standardized format, which can be quickly parsed and processed by both parties. This efficient messaging system is key to the protocol's ability to support high-frequency, real-time communication scenarios.
```

Comment:
An Illustration would've been helpful here. Because you are talking about sockets and connection it would be easier for the reader to understand through a diagram.

Reply:
That's is a great idea, i'll add an illustration.